File: test_files/python/complex_autocomplete.py
"""Complex Python fixture requiring cross-function understanding."""

from dataclasses import dataclass


@dataclass
class Item:
    sku: str
    price: float
    quantity: int


@dataclass
class Coupon:
    code: str
    kind: str  # "percent" or "fixed"
    amount: float


def subtotal(items: list[Item]) -> float:
    return sum(item.price * item.quantity for item in items)


def coupon_discount(current_subtotal: float, coupon: Coupon | None) -> float:
    if coupon is None:
        return 0.0
    if coupon.kind == "percent":
        return min(current_subtotal, current_subtotal * (coupon.amount / 100.0))
    if coupon.kind == "fixed":
        return min(current_subtotal, coupon.amount)
    raise ValueError(f"Unknown coupon kind: {coupon.kind}")


def shipping_cost(after_discount_total: float) -> float:
    if after_discount_total >= 100:
        return 0.0
    if after_discount_total >= 50:
        return 4.99
    return 9.99


def total_with_tax_and_shipping(items: list[Item], coupon: Coupon | None, tax_rate: float) -> float:
    raw_subtotal = subtotal(items)
    discount = coupon_discount(raw_subtotal, coupon)
    discounted_total = raw_subtotal - discount
    shipping = shipping_cost(discounted_total)
    tax = discounted_total * tax_rate
    return round(discounted_total + shipping + tax, 2)


def checkout_preview() -> float:
    basket = [
        Item("keyboard", 80.0, 1),
        Item("cable", 10.0, 2),
    ]
    coupon = Coupon(code="WELCOME10", kind="percent", amount=10)

    return total_with_tax_and_shipping(


def checkout_preview_with_explicit_rate() -> float:
    basket = [
        Item("keyboard", 80.0, 1),
        Item("cable", 10.0, 2),
    ]
    coupon = Coupon(code="WELCOME10", kind="percent", amount=10)
    return total_with_tax_and_shipping(basket, coupon, 

File: test_files/java/ComplexAutocomplete.java
import java.util.List;

public class ComplexAutocomplete {
    record Item(String sku, double price, int quantity) {}
    record Coupon(String code, String kind, double amount) {}

    static double subtotal(List<Item> items) {
        return items.stream().mapToDouble(item -> item.price() * item.quantity()).sum();
    }

    static double couponDiscount(double subtotal, Coupon coupon) {
        if (coupon == null) return 0.0;
        return switch (coupon.kind()) {
            case "percent" -> Math.min(subtotal, subtotal * (coupon.amount() / 100.0));
            case "fixed" -> Math.min(subtotal, coupon.amount());
            default -> throw new IllegalArgumentException("Unknown coupon type: " + coupon.kind());
        };
    }

    static double shipping(double afterDiscount) {
        if (afterDiscount >= 100.0) return 0.0;
        if (afterDiscount >= 50.0) return 5.0;
        return 9.0;
    }

    static double checkoutTotal(List<Item> items, Coupon coupon, double taxRate) {
        double rawSubtotal = subtotal(items);
        double discount = couponDiscount(rawSubtotal, coupon);
        double discounted = rawSubtotal - discount;
        double tax = discounted * taxRate;
        return Math.round((discounted + shipping(discounted) + tax) * 100.0) / 100.0;
    }

    public static void main(String[] args) {
        List<Item> cart = List.of(
            new Item("keyboard", 80.0, 1),
            new Item("cable", 10.0, 2)
        );
        Coupon coupon = new Coupon("WELCOME10", "percent", 10.0);
        Coupon draftCoupon = new Coupon(

        double total = checkoutTotal(
    }
}
